torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,246 @@
+
+
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Modules that lazily initialize parameters, or "lazy modules",
+    derive the shapes of their parameters from the first input(s)
+    to their forward method. Until that first forward they contain
+    :class:torch.nn.UninitializedParameters that should not be accessed
+    or used, and afterward they contain regular :class:torch.nn.Parameters.
+    Lazy modules are convenient since they don't require computing some
+    module arguments, like the in_features argument of a typical :class:torch.nn.Linear.
+
+    After construction, networks with lazy modules should first
+    be converted to the desired dtype and placed on the desired device.
+    The lazy modules should then be initialized with one or more "dry runs".
+    These "dry runs" send inputs of the correct size, dtype, and device through
+    the network and to each one of its lazy modules. After this the network can be used as usual.
+
+    >>> class LazyMLP(torch.nn.Module):
+    ...    def __init__(self):
+    ...        super().__init__()
+    ...        self.fc1 = torch.nn.LazyLinear(10)
+    ...        self.relu1 = torch.nn.ReLU()
+    ...        self.fc2 = torch.nn.LazyLinear(1)
+    ...        self.relu2 = torch.nn.ReLU()
+    ...
+    ...    def forward(self, input):
+    ...        x = self.relu1(self.fc1(input))
+    ...        y = self.relu2(self.fc2(x))
+    ...        return y
+    >>> # constructs a network with lazy modules
+    >>> lazy_mlp = LazyMLP()
+    >>> # transforms the network's device and dtype
+    >>> # NOTE: these transforms can and should be applied after construction and before any 'dry runs'
+    >>> lazy_mlp = mlp.cuda().double()
+    >>> lazy_mlp
+    LazyMLP(
+      (fc1): LazyLinear(in_features=0, out_features=10, bias=True)
+      (relu1): ReLU()
+      (fc2): LazyLinear(in_features=0, out_features=1, bias=True)
+      (relu2): ReLU()
+    )
+    >>> # performs a dry run to initialize the network's lazy modules
+    >>> lazy_mlp(torch.ones(10,10).cuda())
+    >>> # after initialization, LazyLinear modules become regular Linear modules
+    >>> lazy_mlp
+    LazyMLP(
+      (fc1): Linear(in_features=10, out_features=10, bias=True)
+      (relu1): ReLU()
+      (fc2): Linear(in_features=10, out_features=1, bias=True)
+      (relu2): ReLU()
+    )
+    >>> # attaches an optimizer, since parameters can now be used as usual
+    >>> optim = torch.optim.SGD(mlp.parameters(), lr=0.01)
+
+    Note that lazy modules cannot validate if the shape is correct during
+    deserialization when loading an initialized parameter into an uninitialized one.
+    Also note that the initial weights values may become different
+    from the original (non-lazy) module even if the random seed is manually
+    configured, as the order of initialization is different from the original
+    one.
+
+    In the example above, if both layers are `LazyLinear` the initialization
```

Comment:
Now is the bit on initialization. You could do something like this:

"A final caveat when using lazy modules is that the order of initialization of a network's parameters may change, since the lazy modules are always initialized after other modules. This can cause the parameters of a network using lazy modules to be initialized differently than the parameters of a network without lazy modules. For example, if the LazyMLP class defined above had a :class:\`torch.nn.LazyLinear\` module first and then a regular :class:\`torch.nn.Linear\` second, the second module would be initialized on construction and the first module would be initialized during the first dry run."